### PR TITLE
Upgrade OpenSearch version to 1.1.0

### PR DESCRIPTION
**Pull request by cliu123**
##  opensearch-security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 Maintenance
 
 
 2. __Github Issue # or road-map entry, if available:__

 3. __Description of changes:__

 4. __Why these changes are required?__ 

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 

- Built security plugin with OpenSearch 1.1.0
- Installed security plugin on OpenSearch 1.1.0
- [CD run](https://github.com/cliu123/security/runs/3406281257?check_suite_focus=true)

 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by afazel — DISMISSED** [review]

**Review by hsiang9431-amzn — DISMISSED** [review]

**Review by vrozov — CHANGES_REQUESTED** [review]
there are few PR that needs to be merged prior to upgrade

**Comment by dblock** [comment]
What's the latest on merging this? It's blocking https://github.com/opensearch-project/opensearch-build/issues/179.

**Comment by vrozov** [comment]
Waiting for #1224 and #1300.

**Comment by dblock** [comment]
You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.

**Comment by vrozov** [comment]
> You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.

Yes, that change is missing and also change to CD. @cliu123 @afazel @hsiang9431-amzn How much testing was done prior to submitting and approving the PR?

**Comment by cliu123** [comment]
> > You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.
> 
> Yes, that change is missing and also change to CD. @cliu123 @afazel @hsiang9431-amzn How much testing was done prior to submitting and approving the PR?

The opensearch version needs to be changed, but should the security plugin version been changed to `1.1.0.0` right before the release `1.0`?

**Comment by vrozov** [comment]
> > > You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.
> > 
> > 
> > Yes, that change is missing and also change to CD. @cliu123 @afazel @hsiang9431-amzn How much testing was done prior to submitting and approving the PR?

Still need a response on the original question regarding testing/verification of changes.

> The opensearch version needs to be changed, but shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.0`?

@cliu123 No, please check git history of plugin version upgrades and ES/OS upgrades.

**Comment by dblock** [comment]
> > > You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.
> > 
> > 
> > Yes, that change is missing and also change to CD. @cliu123 @afazel @hsiang9431-amzn How much testing was done prior to submitting and approving the PR?
> 
> The opensearch version needs to be changed, but shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.0`?

What do you mean? OpenSearch 1.0 has been released.

**Comment by cliu123** [comment]
> > > > You'll also need https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8b, or the plugin will not install on OpenSearch 1.1.
> > > 
> > > 
> > > Yes, that change is missing and also change to CD. @cliu123 @afazel @hsiang9431-amzn How much testing was done prior to submitting and approving the PR?
> > 
> > 
> > The opensearch version needs to be changed, but shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.0`?
> 
> What do you mean? OpenSearch 1.0 has been released.

@dblock Sorry for confusion! I mean: shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.1`?

**Comment by cliu123** [comment]
> Still need a response on the original question regarding testing/verification of changes.

@vrozov Tested building security plugin with OpenSearch `1.1.0` artifacts.
In order to install the security plugin on OpenSearch `1.1.0`, the opensearch version at [here](https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8bL25) needs to be changed, but I still don't see why changing [OS/ES version](https://github.com/opensearch-project/security/pull/1409/files#diff-38537192da3d6570b9921505e80c780169c9263a1290b506b251c5c387b4ca8bL6) requires to change security plugin version, for [example](https://github.com/opensearch-project/security/pull/706/files)

**Comment by dblock** [comment]
> @dblock Sorry for confusion! I mean: shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.1`?

No, it should be changed to 1.1 right after a release of 1.0, otherwise you keep building 1.0 but it will never be 1.0 anymore.

**Comment by cliu123** [comment]
> > @dblock Sorry for confusion! I mean: shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.1`?
> 
> No, it should be changed to 1.1 right after a release of 1.0, otherwise you keep building 1.0 but it will never be 1.0 anymore.

@dblock  There are 2 different understanding(release processes) here. I believe you are talking about the release process(branch strategy) of OpenSearch core where there is a `1.0` branch, and `main` branch is `1.1`. Security plugin will cut a branch `1.0` soon because some of them open PR to be merged only support OpenSearch `1.1`, and these PRs will be merged in to `main` after creating `1.0` branch. But, on the other hand, security plugin inherits the original release process(branch strategy) from ODFE where plugin versions are upgraded to `1.1` right before `1.1` release. @vrozov Please correct me if I'm wrong or there is anything to be fleshed out.

**Comment by dblock** [comment]
> > > @dblock Sorry for confusion! I mean: shouldn't the security plugin version been changed to `1.1.0.0` right before the release `1.1`?
> > 
> > 
> > No, it should be changed to 1.1 right after a release of 1.0, otherwise you keep building 1.0 but it will never be 1.0 anymore.
> 
> @dblock There are 2 different understanding(release processes) here. I believe you are talking about the release process(branch strategy) of OpenSearch core where there is a `1.0` branch, and `main` branch is `1.1`. Security plugin will cut a branch `1.0` soon because some of the open PRs to be merged only support OpenSearch `1.1`, and these PRs will be merged into `main` after creating `1.0` branch. But, on the other hand, security plugin inherits the original release process(branch strategy) from ODFE where plugin versions are upgraded to `1.1` right before `1.1` release.
> @vrozov Please correct me if I'm wrong or there is anything to be fleshed out. In addidtion, I've reopened the [issue](https://github.com/opensearch-project/security/issues/1296) for branching strategy. Is it possible to elaborate the branching strategy and the reasons behind it in this public Github issue to aviod any more confusion in the future?

Branching != versioning, but they are related.

My rule of thumb is that you create a branch as soon as you need to work on 2 separate versions in parallel, with `main` being the furthermost release. So right now if security's most future release is 1.1 I would expect main to say 1.1. If you are also working on 2.0, then I would say main=2.0, 1.x=1.1 (and 1.0 has shipped and it's 1.0). 

For versioning you want the code to say the version that you're targeting. So when 1.0 release was about to be ready, you may have already started working on 1.1. In that case you would have created a 1.0 branch at that moment, and incremented main to 1.1 to avoid confusion. [Here's what happens if you don't](https://github.com/opensearch-project/OpenSearch/issues/1104).

**Comment by cliu123** [comment]
@dblock @vrozov Upgrading security plugin version to `1.0.1.0` in the [PR#1418](https://github.com/opensearch-project/security/pull/1418). This PR will be merged after the patch release `1.0.1`.

**Comment by vrozov** [comment]
The key question here is whether or not there is a parallel development on multiple versions (2.x, 1.1.x, and 1.0.x). We are currently not working on 2.x and/or 1.1.x. There are several PRs that were planned for 1.0.x and this is what `main` branch targets.

A standard approach for versioning is to use `SNAPSHOT` during development, so the upgrade scheme follows `x.y.z-SNAPSHOT`-> `x.y.z` -> `x.y.z+1-SNAPSHOT`. Note that version change to `x.y.z` is done for the release commit/tag and then the version is incremented and is changed back to `SNAPSHOT`. Due to the lack of `SNAPSHOT` support for the security plugin, the version is changed at the release time as it pretty much does not matter when the version is changed: commits that follow the release commit is neither `x.y.z` nor `x.y.z+1` version.

**Comment by cliu123** [comment]
@vrozov @dblock I've create the [branch](https://github.com/opensearch-project/security/tree/1.0) `1.0` at [the current latest commit](https://github.com/opensearch-project/security/commit/1d443fb7f7c0eb42de0b95e72c11141eb6d17f72) on `main` branch.

**Review by vrozov — APPROVED** [review]
@cliu123 no need to create a new branch for now. Tag the previous commit as `v1.0.1-rc1-OS`.

**Review by afazel — APPROVED** [review]

File: .github/workflows/cd.yml [hunk 1]
```diff
@@ -31,7 +31,7 @@ jobs:
       with:
         repository: 'opensearch-project/OpenSearch'
         path: OpenSearch
-        ref: '1.0'
```
Comment (cliu123) [inline .github/workflows/cd.yml:34]: Would you please also change in [CD](https://github.com/opensearch-project/security/blob/main/.github/workflows/cd.yml#L38)?

**Pull request merged**